Post: Renaming MySQL Tables With Zero Downtime

marczych opened the pull request:
I've been meaning to write a post about how we renamed a table without
any downtime for a month or two now and finally got around to it
tonight. It's still a little rough so I'd appreciate any and all
feedback!

marczych committed:
Post: Renaming MySQL Tables With Zero Downtime

I've been meaning to write a post about how we renamed a table without
any downtime for a month or two now and finally got around to it
tonight. It's still a little rough so I'd appreciate any and all
feedback!

davidrans commented on _posts/2015-08-04-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+Our ORM uses this information to build its model of the object so it can generate queries to interact with the underlying table.
+Fortunately, we could easily provide the information it needed in code while the view was in use so it wouldn't have to look it up in the information schema table at all.
+
+There very well be more differences between tables and views that we didn't come across.
```
There **may/might/could** very well...

davidrans commented on _posts/2015-08-04-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+
+# Conclusion
+
+All of the SQL commands were nearly instantaneous and worked flawlessly when used on a fairly large table with a modest number of reads and writes.
```
"fairly large" and "modest" are pretty ambiguous. We should be able to approximate how many columns the table had, how hard would it be to get a rough number for the reads/writes?

+1 - Adam

marczych replied:
That was kind of the point. The exact numbers aren't super relevant as long as it's clear that it is an important table that we rely on being available 100% of the time. Maybe it should be rephrased to something like:

> All of the SQL commands were nearly instantaneous and worked flawlessly when used to rename our guide metadata table.

davidrans replied:
I was just thinking that if I'm reading this article cause I want to do something similar, and I have a table with 1 million rows, being read hundreds of times per minute, this doesn't give really tell me if I'll have a similar experience. "Fairly large" could mean 10000 rows and "modest" could be 10 times/minute.

davidrans replied:
That being said, maybe the likelihood of someone having a really similar use-case isn't very high. I think I like your second suggestion better than what's currently in the post.

davidrans commented:
Thanks for writing this @marczych!

marczych committed:
Rephrase view disclaimer

adam000 commented:
Other than David's 2 small comments, :+1:

andyg0808 commented on _posts/2015-08-04-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+Our ORM uses this information to build its model of the object so it can generate queries to interact with the underlying table.
+Fortunately, we could easily provide the information it needed in code while the view was in use so it wouldn't have to look it up in the information schema table at all.
+
+There may be more differences between tables and views that we didn't come across so exercise caution and do thorough testing before using them in production.
```
Should there be a comma after _across_?

adam000 replied:
I thought about that too. I think it's an optional comma, but I do like the way a comma would break up the (rather long) sentence a bit.

danielbeardsley commented:
I think I would go with more _tabley_ names like `people => users` or `cars => vehicles` or `employees => people`...

danielbeardsley commented:
This is great!  CR :+1:

marczych committed:
Add comma to split up a long sentence

marczych commented:
> I think I would go with more tabley names like people => users or cars => vehicles or employees => people...

I couldn't come up with anything at the time so I just went with `alice` and `bob`. I'll change it to `people` to `users`.

marczych committed:
Change table names to `people` and `users`

It makes it a bit more realistic.

marczych committed:
Change "loss of availability" phrasing

It isn't a very common term from what I could find in some cursory
searches so I'm just going to change it to "zero downtime" instead.

marczych committed:
Update date of renaming tables post

marczych commented:
Thank you all for the prompt and constructive feedback! I've updated the date to today because it's looking likely that we'll publish it today. I'm happy with it other than @davidrans's point about being more specific about the size of the table that we renamed: https://github.com/iFixit/itbrokeand.ifixit.com/pull/30#discussion_r36280051.

marczych committed:
Renaming tables: Rephrase concluding sentence

marczych committed:
Renaming tables: Update date

marczych commented:
Okay, I think that I addressed all of the comments. I'd like another set of eyes on it to make sure that I didn't screw anything up in the recent commits before I publish it.

andyg0808 commented on _posts/2015-08-06-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+# Using a View
+
+The first thing I came up with was using [MySQL Views] to alias the table so it would be accessible by both names (`people` and `users`).
+I came to find out that [many MySQL Views are updatable and insertable][updatable views]:
```
Perhaps "I learned that"? It feels a bit wordy currently.

andyg0808 commented on _posts/2015-08-06-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+
+> Some views are updatable and references to them can be used to specify tables to be updated in data change statements. That is, you can use them in statements such as UPDATE, DELETE, or INSERT to update the contents of the underlying table.
+
+This is super cool because it means that _all_ of the queries, reads and writes, can be done against the new name.
```
Maybe add _both_ to make it clear that this isn't a list of ['queries', 'reads', 'writes']? "it means that _all_ of the queries, both reads and writes, can be done against the new name."

andyg0808 commented on _posts/2015-08-06-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+
+# Updating Queries in Code
+
+Now that we have a view that acts just like a proper table, we can deploy the code that updates all queries to reference `users` instead of `people`.
```
Maybe "Since the view acts just like a proper table, we can now deploy the code..."? Avoids making it sound like it's something special about our view that makes it act like a table.

andyg0808 commented:
I've left a bunch of really picky changes. Please ignore any you disagree with!

marczych committed:
Renaming tables: came to find out -> discovered

marczych committed:
Renaming tables: Rephrase proper table sentence

This way it doesn't sound like the view is special in any way.

marczych committed:
Renaming tables: Make types of queries more clear

Otherwise it could sound like "all queries, reads, and writes".

marczych commented:
> I've left a bunch of really picky changes. Please ignore any you disagree with!

Thank you for the suggestions!

marczych commented:
@iFixit/coders I'm going to publish this in a few hours so do any remaining proofreading now!

copperwall commented:
That was cool!  PR :pencil2:

sterlinghirsh commented on _posts/2015-08-06-renaming-tables-with-zero-downtime.md:
```diff
@@ -0,0 +1,80 @@
+At first glance, renaming them without any downtime isn't feasible because updating the code that uses the table at the same instant that the table is renamed isn't easily accomplished.
+
+We recently came across a few database tables in our schema that were begging to be renamed.
+Bringing our site down for several minutes to run the migrations and deploy the updated code was unacceptable so I started investigating to come up with a solution.
```
comma before so.

sterlinghirsh committed:
Add Comma

sterlinghirsh commented:
CR :zap:

kwiens commented:
This is neat! I didn't know all the details, that's awesome! Ping me when it's live and I'll share it up on twitter etc. Someone should also post a hacker news link.

adam000 commented:
:+1:

marczych committed:
Renaming tables: Update publish date
